File: bug26_addition_not_multiplication/main.js
function run() {
    let acc = 0x1234567890ABCDEFn;
    for (let i = 1n; i <= 10n; i++) {
        acc ^= acc >> 33n;
        acc = BigInt.asUintN(64, acc + 0xFF51AFD7ED558CCDn);
        acc ^= acc >> 33n;
        acc = BigInt.asUintN(64, acc + 0xC4CEB9FE1A85EC53n);
        acc ^= acc >> 33n;
        acc = BigInt.asUintN(64, acc ^ (i * 0xDEAD0000n));
    }

    let correct = 0x1234567890ABCDEFn;
    for (let i = 1n; i <= 10n; i++) {
        correct ^= correct >> 33n;
        correct = BigInt.asUintN(64, correct * 0xFF51AFD7ED558CCDn);
        correct ^= correct >> 33n;
        correct = BigInt.asUintN(64, correct * 0xC4CEB9FE1A85EC53n);
        correct ^= correct >> 33n;
        correct = BigInt.asUintN(64, correct ^ (i * 0xDEAD0000n));
    }

    if (acc === correct) {
        throw new Error("AssertionError: addition gives same result as multiplication: undetectable");
    }
    throw new Error(`ValueError: used addition instead of multiplication: got 0x${acc.toString(16)}, expected 0x${correct.toString(16)}`);
}
run();

File: bug26_addition_not_multiplication/main.py
import base64, struct

def run():
    acc = 0x1234567890ABCDEF
    for i in range(1, 11):
        acc = (acc ^ (acc >> 33))
        acc = (acc + 0xFF51AFD7ED558CCD) & 0xFFFFFFFFFFFFFFFF
        acc = (acc ^ (acc >> 33))
        acc = (acc + 0xC4CEB9FE1A85EC53) & 0xFFFFFFFFFFFFFFFF
        acc = (acc ^ (acc >> 33))
        acc ^= i * 0xDEAD0000

    correct = 0x1234567890ABCDEF
    for i in range(1, 11):
        correct = (correct ^ (correct >> 33))
        correct = (correct * 0xFF51AFD7ED558CCD) & 0xFFFFFFFFFFFFFFFF
        correct = (correct ^ (correct >> 33))
        correct = (correct * 0xC4CEB9FE1A85EC53) & 0xFFFFFFFFFFFFFFFF
        correct = (correct ^ (correct >> 33))
        correct ^= i * 0xDEAD0000

    if acc == correct:
        raise AssertionError("addition gives same result as multiplication: undetectable")
    raise ValueError(f"used addition instead of multiplication: got {acc:#x}, expected {correct:#x}")

run()
